src/cascadia/TerminalControl/ThrottledFunc.h

```diff
@@ -4,66 +4,88 @@ Licensed under the MIT license.
 
 Module Name:
 - ThrottledFunc.h
-
-Abstract:
-- This module defines a class to throttle function calls.
-- You create an instance of a `ThrottledFunc` with a function and the delay
-  between two function calls.
-- The function takes an argument of type `T`, the template argument of
-  `ThrottledFunc`.
-- Use the `Run` method to wait and then call the function.
 --*/
 
 #pragma once
 #include "pch.h"
 
 #include "ThreadSafeOptional.h"
 
+// Class Description:
+// - Represents a function whose invocation is delayed by a specified duration
+//   and rate-limited such that if the code tries to run the function while a
+//   call to the function is already pending, the request will be ignored.
+// - The function will be run on the the specified dispatcher.
+class ThrottledFunc : public std::enable_shared_from_this<ThrottledFunc>
```

Comment:
I have an idea. Maybe I could have a class `Arg<T>` and another class `NoArg` that contain the code that is different between the two cases (for when there is an argument, with a `ThreadSafeOptional` and when there no argument, with a `std::atomic_flag`) that I would then use in `ThrottleFunc<Arg<T>>` or `ThrottleFunc<NoArg>`. Would it be better this way?

Reply:
We could use rest arguments for this like so:
```
template<typename ...Args>
struct ThrottledFunc : public std::enable_shared_from_this<ThrottledFunc<Args...>> {
    using Func = std::function<void(Args...)>;

    ThrottledFunc(Func func) : _func(std::move(func)) {}

    void Run(Args&& ...args) {
        _pendingRunArgs.emplace(std::forward<Args>(args)...); // store
        std::apply(_func, _pendingRunArgs.value()); // call
    }

private:
    Func _func;
    std::optional<std::tuple<Args...>> _pendingRunArgs;
};
```

Use it like this:
```
auto a = std::make_shared<ThrottledFunc<int>>([](int x) {});
a->Run(123);

auto b = std::make_shared<ThrottledFunc<>>([]() {});
b->Run();
```

You can probably make it work as `ThrottledFunc<void(int)>` too for better optics, but I'm personally not sure how one would do that.

Reply:
Yes but for void() I want to use std::atomic_flag instead of an std::optional<std::tuple<>> with a std::mutex because it's faster.

Reply:
Did you measure the impact of using an `atomic_flag` as opposed to a `mutex` though?
I'm fairly certain the difference is smaller than the deviation of potential benchmarks to be honest. 🙈
I can extend the above template to support flags though. Makes it a bit larger, but shouldn't really affect compile time. Would you like me to do so?

Reply:
I've thought about it and I honestly believe it's better to not keep it DRY if you want to specialise the `void()` case. I'd suggest implementing it with a template specialization though instead of having 2 seperate classes. I.e. using:
```cpp
template<>
class ThrottledFunc<void> {
    // ...
}
```

(If you implement it using variadic templates you need to use `<>` for the specialization though.)

Reply:
BTW I really only intend these comments as suggestions. 🙈 It's absolutely fine if you want to do it any other way. 👍

Reply:
Do you know if there is a better way to do this: https://github.com/microsoft/terminal/pull/6492/commits/92abce305710f303cadf226875167a585284052f#diff-7b52efc3278e2ed0f27eb5ba4397fc7fR64-R65?
The ThrottledArgsFunc class has the template argument Args, but the Run function cannot be Run(Args&&... args) because the forward reference thing doesn't work here because Args may not be a reference (for example, if Args=int, then we get Run(int&&) which means that Run(lvalue) won't work).

Reply:
Mutex is 3% of CPU:
![atomic_flag](https://user-images.githubusercontent.com/56923875/84708451-d1cfd000-af60-11ea-802a-7ecc1d08717f.png)
![mutex](https://user-images.githubusercontent.com/56923875/84708453-d300fd00-af60-11ea-9501-fd08119b93bf.png)

Reply:
I tried to do some DRY with my idea above with `Args` and `NoArgs` but I don't like it. The code becomes a monster 😆.

Reply:
I believe you did it perfectly. `std::optional` does it the same way.
If you want to further optimize it, the only thing I'd change is to move as much as possible (but at least the `_dispatcher.RunAsync` part) into a seperate function. That way there's only 1 template instantiation of that function per `Args...`, instead of 1 per `MakeArgs...`.

Reply:
Ok. I think I won't do this because I don't have a good name for that separate function so it would make the code too confusing.

Reply:
Wow that's a lot! I suppose because on Windows mutexes are implemented using critical sections whose implementation is in a .dll? That's crazy! In a sad way... 😞 I can't imagine this to be the case with Rust or Go tbh...

As mentioned before you could use template specialization to unify the classname at least: https://gist.github.com/lhecker/8b465dd8232669593c1e60efe25e05d3